nni/algorithms/compression/v2/pytorch/pruning/basic_pruner.py

```diff
@@ -0,0 +1,506 @@
+                        loss.backward()
+                        optimizer.step()
+        optimizer
+            The optimizer instance used in trainer. Note that this optimizer might be patched during collect data,
+            so do not use this optimizer in other places.
+        criterion
+            The criterion function used in trainer. Take model output and target value as input, and return the loss.
+        training_batches
+            The batch number used to collect activations.
+        mode
+            'normal', 'dependency_aware' or 'global'.
+
+            If prune the model in a dependency-aware way, this pruner will
+            prune the model according to the l1-norm of weights and the channel-dependency or
+            group-dependency of the model. In this way, the pruner will force the conv layers
+            that have dependencies to prune the same channels, so the speedup module can better
+            harvest the speed benefit from the pruned model. Note that, if set 'dependency_aware'
+            , the dummy_input cannot be None, because the pruner needs a dummy input to trace the
+            dependency between the conv layers.
+
+            If prune the model in a global way, all layer weights with same config will be considered uniformly.
+            That means a single layer may not reach or exceed the sparsity setting in config,
+            but the total pruned weights meet the sparsity setting.
+        max_sparsity_per_layer
+            The max sparsity per layer constraint, to prevent all weight in a layer be pruned in global mode.
+        dummy_input
+            The dummy input to analyze the topology constraints. Note that, the dummy_input
+            should on the same device with the model.
+        """
+        self.mode = mode
+        self.max_sparsity_per_layer = max_sparsity_per_layer
+        self.dummy_input = dummy_input
+        self.trainer = trainer
+        self.optimizer = optimizer
+        self.criterion = criterion
+        self.training_batches = training_batches
+        super().__init__(model, config_list)
+
+    def validate_config(self, model: Module, config_list: List[Dict]):
+        schema = PrunerSchema([{
+            SchemaOptional('sparsity'): And(float, lambda n: 0 < n < 1),
+            'op_types': ['Conv2d'],
+            SchemaOptional('op_names'): [str],
+            SchemaOptional('exclude'): bool
+        }], model, _logger)
+
+        schema.validate(config_list)
+
+    def _collector(self, buffer: List, weight_tensor: Tensor) -> Callable[[Module, Tensor, Tensor], None]:
+        def collect_taylor(grad: Tensor):
+            if len(buffer) < self.training_batches:
+                buffer.append(self._calculate_taylor_expansion(weight_tensor, grad))
+        return collect_taylor
+
+    def _calculate_taylor_expansion(self, weight_tensor: Tensor, grad: Tensor) -> Tensor:
+        return (weight_tensor * grad).data.pow(2)
```

Comment:
Do we need `detach()` here?

Reply:
what will happen if we don't `detach()` here

Reply:
I think if some operations temper with the data the weight might be affected, but it seems that this chance is very small. In our previous implementation, we also used `data()` e.g., in the `get_channel_sum` function in the masker to get the data for score calculation. However, the masker immediately `detach()` the resulting mask, and does not expose the intermediate data. Here since we modularize each step and pass the data around to other components, it might be safer to `detach()`, but I'm also not sure.

Reply:
yes, as we discussed, detach() is safer, add it